### PR TITLE
ENH: Moving Ridge out of preview

### PR DESCRIPTION
**Pull request by DDJHB**
## Description

This PR moves Ridge out of preview namespace to main sklearnex. It drives sklearnex Ridge away from daal4py by switching underlying implementation to onedal4py, which enables GPU support in turn. Minor usage of daal4py remains in case of individual (non-scalar) penalties since oneDAL still does not support such scenario. Performance comparison has been recorded on CPU:
image
Performance comparison on GPU is irrelevant since there was not real GPU implementation before. 
Separate ticket to update documentation has been created.

---

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by DDJHB** [comment]
Private CI errors seem unrelated

File: sklearnex/linear_model/tests/test_ridge.py [hunk 1]
```diff
@@ -37,7 +41,7 @@ def test_sklearnex_import_ridge(dataframe, queue):
     ridge_reg = Ridge(alpha=0.5).fit(X_c, y_c)
 
     if daal_check_version((2024, "P", 600)):
-        assert "preview" in ridge_reg.__module__
+        assert "sklearnex" in ridge_reg.__module__
```
Comment (ethanglaser) [inline sklearnex/linear_model/tests/test_ridge.py:44]: Would it make sense to also assert that preview is not in the module string?

**Comment by ethanglaser** [comment]
This should also be updated - https://github.com/intel/scikit-learn-intelex/blob/main/doc/sources/preview.rst?plain=1#L49-L76, remove ridge from bottom part and replace it with EmpiricalCovariance or IncrementalPCA in snippet

File: sklearnex/tests/test_n_jobs_support.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def _check_n_jobs_entry_in_logs(records, function_name, n_jobs):
         if f"{function_name}: setting {expected_n_jobs} threads" in rec:
             return True
     # False if n_jobs is set and not found in logs
-    return n_jobs is None
+    return n_jobs is None or n_jobs == -1
```
Comment (ethanglaser) [inline sklearnex/tests/test_n_jobs_support.py:52]: What is the reason for this?

File: sklearnex/linear_model/ridge.py [hunk 3]
```diff
@@ -14,11 +14,416 @@
...
+            self.n_features_in_ = self._onedal_estimator.n_features_in_
+            self._coef = self._onedal_estimator.coef_
+            self._intercept = self._onedal_estimator.intercept_
+
+        fit.__doc__ = _sklearn_Ridge.fit.__doc__
+        predict.__doc__ = _sklearn_Ridge.predict.__doc__
+        score.__doc__ = _sklearn_Ridge.score.__doc__
+
+else:
+    from daal4py.sklearn.linear_model._ridge import Ridge
```
Comment (ethanglaser) [inline sklearnex/linear_model/ridge.py:420]:
```suggestion
    from daal4py.sklearn.linear_model import Ridge

```

**Review by ethanglaser — COMMENTED** [review]
Looking great!

File: onedal/linear_model/linear_model.py [hunk 4]
```diff
@@ -293,15 +293,15 @@ def fit(self, X, y, queue=None):
         """
         module = self._get_backend("linear_model", "regression")
 
-        X = _check_array(
```
Comment (ethanglaser) [inline onedal/linear_model/linear_model.py:296]: Is this what was causing the slowdown between sklearnex and daal4py?

File: sklearnex/tests/test_n_jobs_support.py (hunk 2 above)
Comment (DDJHB) [reply sklearnex/tests/test_n_jobs_support.py:52]: this is a small bug i discovered with the n_jobs_support test - when you run it in batch with other tests (what happens in CI) it works perfectly, yet, run standalone - it fails on n_jobs=-1. upon little investigation found following:
n_jobs_wrapper in daal4py does not log anything since for -1 case it matches with old thread count which means nothing is to be set so nothing is rly logged, which fails the test since it relies on logs to pass. agreed with @icfaust to add this clause so whenever we have n_jobs=-1 and "return True" on line 50 did not trigger, meaning that no update to n_threads has been done, success is still returned for the check in the test to pass.

**Comment by icfaust** [comment]
With the CI fixes merged, please rebase @DDJHB

File: doc/sources/preview.rst [hunk 5]
```diff
@@ -49,11 +49,11 @@ For example, you can set the environment variable in the following way:
 Then, you can import Scikit-learn estimator patched with a preview one from `sklearnex.preview` module::
 
      from sklearnex import patch_sklearn
-     patch_sklearn()
-     from sklearn.linear_model import Ridge
-     print(Ridge.__module__)
+     patch_sklearn(preview=True)
```
Comment (ethanglaser) [inline doc/sources/preview.rst:52]:
```suggestion
     patch_sklearn()
```
arg should not needed due to env variable setting above

File: sklearnex/linear_model/tests/test_ridge.py
Comment (ethanglaser) [inline sklearnex/linear_model/tests/test_ridge.py:1]: Requires linting

File: doc/sources/preview.rst (hunk 5 above)
Comment (DDJHB) [reply doc/sources/preview.rst:52]: yeah, did not notice env var setting in the file. quick question, why show customer a way to enable preview by setting env variable instead of explicit indication in code?

File: onedal/linear_model/linear_model.py [hunk 6]
```diff
@@ -293,15 +293,15 @@ def fit(self, X, y, queue=None):
...
-            X,
-            dtype=[np.float64, np.float32],
-            force_all_finite=False,
-            ensure_2d=False,
-            copy=self.copy_X,
-        )
+        if not isinstance(X, np.ndarray):
```
Comment (icfaust) [inline onedal/linear_model/linear_model.py:296]: @samir-nasibli  Is there a ticket for Ridge Array_api support? I know you had mentioned you wanted this to be more in the direction of array_api, but I think maybe it should be split into 2 PRs - out of preview and array_api support.  If thats okay, could you make sure to send @DDJHB and me the ticket, or we can have @DDJHB make it.  You can see this code is rather numpy-oriented.

File: sklearnex/linear_model/ridge.py [hunk 7]
```diff
@@ -14,11 +14,416 @@
...
+
+    from .._device_offload import dispatch, wrap_output_data
+    from .._utils import PatchingConditionsChain
+
+    if sklearn_check_version("1.6"):
+        from sklearn.utils.validation import validate_data
+    else:
+        validate_data = _sklearn_Ridge._validate_data
+
+    def _is_numeric_scalar(value):
```
Comment (icfaust) [inline sklearnex/linear_model/ridge.py:49]: Still not sure why this was/is a separate function...

File: sklearnex/linear_model/ridge.py [hunk 8]
```diff
@@ -14,11 +14,416 @@
...
+                        self.solver == "auto",
+                        f"'{self.solver}' solver is not supported. "
+                        "Only 'auto' solver is supported.",
+                    ),
+                    (
+                        not issparse(X) and not issparse(y),
+                        "Sparse input is not supported.",
+                    ),
+                    (
+                        not is_underdetermined,
```
Comment (icfaust) [inline sklearnex/linear_model/ridge.py:222]: This is worth checking against https://github.com/intel/scikit-learn-intelex/pull/2083 where underdetermined checks are now only necessary for GPU.

File: sklearnex/linear_model/ridge.py [hunk 9]
```diff
@@ -14,11 +14,416 @@
...
+else:
+    from daal4py.sklearn.linear_model import Ridge
+    from onedal._device_offload import support_input_format
+
+    Ridge.fit = support_input_format(queue_param=False)(Ridge.fit)
+    Ridge.predict = support_input_format(queue_param=False)(Ridge.predict)
+    Ridge.score = support_input_format(queue_param=False)(Ridge.score)
+
+    logging.warning(
+        "Preview Ridge requires oneDAL version >= 2024.6 but it was not found"
```
Comment (icfaust) [inline sklearnex/linear_model/ridge.py:428]:
```suggestion
        "Ridge requires oneDAL version >= 2024.6 but it was not found"

```

File: sklearnex/linear_model/tests/test_ridge.py [hunk 10]
```diff
@@ -100,3 +106,15 @@ def test_ridge_predict_before_fit(dataframe, queue):
 
         with pytest.raises(NotFittedError):
             model.predict(X_c)
+
```
Comment (icfaust) [inline sklearnex/linear_model/tests/test_ridge.py:109]: Maybe add some tests like from #2083?

**Review by icfaust — COMMENTED** [review]
Some questions about overdetermined systems, a little bit about a warning.  Finally, we should make Docs green. Close to merge.

File: sklearnex/linear_model/ridge.py (hunk 7 above)
Comment (DDJHB) [reply sklearnex/linear_model/ridge.py:49]: single point of change for expression applied in different parts of the class

Comment (DDJHB) [reply sklearnex/linear_model/ridge.py:49]: resolving as per our conversation @icfaust

File: doc/sources/preview.rst (hunk 5 above)
Comment (ethanglaser) [reply doc/sources/preview.rst]: Good point and I tend to agree on setting using code instead of env variable. Any objections to this @Alexsandruss ?

**Comment by david-cortes-intel** [comment]
> @david-cortes-intel does this PR look properly aligned with #2083 (specifically see changes from commit: [4e18d8a#diff-ed94dee2dafa5407c12f73420aaee87ee7748a2fedeeea8d4bd7b1d50fbf4a53](https://github.com/intel/scikit-learn-intelex/commit/4e18d8a8c5039d911d2c8a4f13d59bb8a13a4c60#diff-ed94dee2dafa5407c12f73420aaee87ee7748a2fedeeea8d4bd7b1d50fbf4a53))

Yes, changes look aligned - I had made a mistake in the condition namings which this PR is correcting.

File: onedal/linear_model/linear_model.py (hunk 6 above)
Comment (icfaust) [reply onedal/linear_model/linear_model.py:296]: @samir-nasibli ping

Comment (samir-nasibli) [reply onedal/linear_model/linear_model.py:296]: Currently only numpy.ndarray array inputs are expected. No need for branching. I think now this branching is for avoiding sparse inputs.

File: onedal/linear_model/linear_model.py [hunk 11]
```diff
@@ -293,15 +293,15 @@ def fit(self, X, y, queue=None):
...
+            X = np.asarray(X)
+
+        dtype = get_dtype(X)
+        if dtype not in [np.float32, np.float64]:
+            dtype = np.float64
+            X = X.astype(dtype, copy=self.copy_X)
 
-        y = np.asarray(y).astype(dtype=get_dtype(X))
+        y = np.asarray(y).astype(dtype=dtype)
```
Comment (samir-nasibli) [inline onedal/linear_model/linear_model.py:304]: This is still unclear to me why we have this perf drop here. How much this impl differ from daal4py CPU branch?

File: sklearnex/linear_model/ridge.py [hunk 12]
```diff
@@ -14,11 +14,433 @@
...
+                super().__init__(
+                    alpha=alpha,
+                    fit_intercept=fit_intercept,
+                    normalize=normalize,
+                    copy_X=copy_X,
+                    max_iter=max_iter,
+                    tol=tol,
+                    solver=solver,
+                    random_state=random_state,
+                )
```
Comment (samir-nasibli) [inline sklearnex/linear_model/ridge.py:138]: For the new estimators or functionality we are not supporting skl < 1.0

File: sklearnex/linear_model/tests/test_ridge.py [hunk 13]
```diff
@@ -100,3 +106,48 @@ def test_ridge_predict_before_fit(dataframe, queue):
...
+
+@pytest.mark.skip(reason="Forced to use original sklearn for now.")
+def test_sklearnex_multivariate_ridge_alpha_shape():
+    from sklearnex.linear_model import Ridge
+
+    _test_multivariate_ridge_alpha_shape(Ridge, random_state=0)
+
+
+if daal_check_version((2025, "P", 100)):
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/tests/test_ridge.py:125]: This one could be changed to `pytest.mark.skipif`.

File: sklearnex/linear_model/tests/test_ridge.py [hunk 14]
```diff
@@ -100,3 +106,48 @@ def test_ridge_predict_before_fit(dataframe, queue):
...
+
+    @pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+    @pytest.mark.parametrize("overdetermined", [True, False])
+    @pytest.mark.parametrize("alpha", [0.1, 0.5, 1.0])
+    def test_ridge_overdetermined_system(dataframe, queue, overdetermined, alpha):
+        if not overdetermined and queue and queue.sycl_device.is_gpu:
+            pytest.skip(reason="Underdetermined system is not yet supported on GPU")
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/tests/test_ridge.py:132]: I'm not 100% sure how ridge is implemented, but if it uses the same code as the non-regularized linear regression, you could just enable this in the fit conditions and it should work (provided the regularization is sufficiently higher than zero for the dtype), because the regularization would make the linear systems positive.

Comment (DDJHB) [reply sklearnex/linear_model/tests/test_ridge.py:132]: this is a very nice catch, thanks

File: doc/sources/preview.rst (hunk 5 above)
Comment (Alexsandruss) [reply doc/sources/preview.rst]: I think it's better to replace env. variable setting with patching call but add note below for variable.

File: sklearnex/linear_model/tests/test_ridge.py [hunk 15]
```diff
@@ -100,3 +106,48 @@ def test_ridge_predict_before_fit(dataframe, queue):
...
+            y = numpy.random.rand(10)
+
+        X_c = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+        y_c = _convert_to_dataframe(y, sycl_queue=queue, target_df=dataframe)
+        ridge_reg = Ridge(alpha=alpha, fit_intercept=False).fit(X_c, y_c)
+
+        # computing the coefficients manually
+        lambda_identity = alpha * numpy.eye(X.shape[1])
+        inverse_term = numpy.linalg.inv(numpy.dot(X.T, X) + lambda_identity)
+        xt_y = numpy.dot(X.T, y)
```
Comment (ethanglaser) [inline sklearnex/linear_model/tests/test_ridge.py:150]: Let's import numpy as np

File: sklearnex/linear_model/ridge.py [hunk 16]
```diff
@@ -236,18 +231,21 @@ def _onedal_gpu_supported(self, method_name, *data):
...
-                patching_status.and_conditions(
-                    [
-                        (
-                            not is_underdetermined,
-                            "The shape of X (fitting) does not satisfy oneDAL requirements:"
-                            "Number of features + 1 >= number of samples.",
-                        ),
-                    ]
-                )
+                if isinstance(self.alpha, numbers.Real) and np.isclose(self.alpha, 0):
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/ridge.py:234]: I don't think `np.isclose` would suffice here. This regularizaation number is meant to be added to a large number which is the result of sums of squares. Therefore, at the very least, it should be larger than machine epsilon for the desired dtype, and likely higher in order to ensure that $ssq + \alpha > ssq$.

Perhaps 1e-5 could be a more reasonable minimum threshold.

File: sklearnex/linear_model/ridge.py [hunk 17]
```diff
@@ -236,18 +231,21 @@ def _onedal_gpu_supported(self, method_name, *data):
...
+                    n_features = _num_features(data[0], fallback_1d=True)
+                    is_underdetermined = n_samples < (
+                        n_features + int(self.fit_intercept)
+                    )
+                    patching_status.and_conditions(
+                        [
+                            (
+                                not is_underdetermined,
+                                "The shape of X (fitting) does not satisfy oneDAL requirements:"
+                                "Number of features + 1 >= number of samples.",
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/ridge.py:245]: Here it could also mention the regularization being non-zero as another requirement.

File: sklearnex/linear_model/ridge.py [hunk 18]
```diff
@@ -14,11 +14,374 @@
...
+
+            return patching_status
+
+        def _onedal_gpu_supported(self, method_name, *data):
+            patching_status = PatchingConditionsChain(
+                f"sklearn.linear_model.{self.__class__.__name__}.{method_name}"
+            )
+
+            if method_name == "fit":
+                if isinstance(self.alpha, numbers.Real) and np.isclose(self.alpha, 0):
```
Comment (david-cortes-intel) [inline sklearnex/linear_model/ridge.py:234]: Also since the PR that fixes this for GPU was already merged (https://github.com/oneapi-src/oneDAL/pull/2976), perhaps you could also skip this check for version >= 2025.2.

**Review by icfaust — COMMENTED** [review]
I hate to be the barer of bad news, please remove the following skips and rerun private CI:
https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/tests/test_patching.py#L147
https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/tests/test_patching.py#L138
https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/tests/test_memory_usage.py#L75

**Review by icfaust — COMMENTED** [review]
@DDJHB Please merge main into your PR and rerun, some conventions in the codebase have changes, and some CI failures have been remedied.

**Review by david-cortes-intel — APPROVED** [review]
LGTM

**Review by icfaust — APPROVED** [review]

**Review by Alexsandruss — APPROVED** [review]
